crosscluster/logical: enable automatic bidi replication

msbutler opened the pull request:
This patch enables a user to create a bidirectional replication stream with one
command. When the user runs `CREATE LOGICALLY REPLICATED TABLE tabB FROM tabA
ON uriA  WITH BIDIRECTIONAL ON uriB`, the following will happen:
- on cluster B, during job planning, construct the reverse stream cmd, while we
	still have access to original LDR cmd. This command will have the following
form: `CREATE LOGICAL REPLICATION STREAM FROM TABLE tabB ON 'uriB' INTO TABLE
tabA WITH CURSOR = $1, PARENT = '{og job id}';`
- on cluster B, create table tabB and begin an offline scan, as normal
- on cluster B, once the offline scan completes, but before tabB is published,
  set up the reverse stream from B to A at a cursor time after the initial scan
completes but before the tabB has come online. This cursor time prevents data
looping and ensures the replication of all future foreground writes.
- on cluster B, steady state replication begins.

Epic: none

Release note (sql change): when a user runs CREATE LOGICALLY REPLICATED TABLE,
they must specify one of the following options:
- UNIDIRECTIONAL: setups a unidirectional stream with fast initial scan
- BIDIRECTIONAL ON {dest uri}: sets up a bidirectional stream from the original
  dest to the original source.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/138297)
<!-- Reviewable:end -->

msbutler committed:
sql/parser: introduce more LDR options for bidi replication

This patch introduces two options to CREATE LOGICALLY REPLICATED TABLES:
- UNIDIRECTIONAL: to specify a unidectional ldr stream with a fast offline scan
- BIDIRECTIONAL ON {uri}: which customers will use to set up a bidirectional
  stream with one command. The uri provided by the customer connects to the
source of the reverse stream, aka the destination of the main stream.

The PARENT option will be used internally to set up the reverse stream.
Specifically, when the destination sets up the reverse stream, it will pass its
jobID as the parent for the reverse stream, which the reverse stream will use
to prevent creating duplicate reverse streams and to configure a stream that
replicates from an offline source. Customers should not directly use it.

Epic: none

Release note: none

msbutler committed:
crosscluster/streamclient: add ExecStatement api

This destination will use this api to set up a reverse stream from a cursor.

Epic: none

Release note: none

msbutler committed:
crosscluster/producer: allow LDR creation from offline tables

This patch allows a user to create an LDR stream that replicates from an
offline table.

Epic: none

Release note: none

msbutler committed:
crosscluster/logical: enable automatic bidi replication

This patch enables a user to create a bidirectional replication stream with one
command. When the user runs `CREATE LOGICALLY REPLICATED TABLE tabB FROM tabA
ON uriA  WITH BIDIRECTIONAL ON uriB`, the following will happen:
- on cluster B, during job planning, construct the reverse stream cmd, while we
	still have access to original LDR cmd. This command will have the following
form: `CREATE LOGICAL REPLICATION STREAM FROM TABLE tabB ON 'uriB' INTO TABLE
tabA WITH CURSOR = $1, PARENT = '{og job id}';`
- on cluster B, create table tabB and begin an offline scan, as normal
- on cluster B, once the offline scan completes, but before tabB is published,
  set up the reverse stream from B to A at a cursor time after the initial scan
completes but before the tabB has come online. This cursor time prevents data
looping and ensures the replication of all future foreground writes.
- on cluster B, steady state replication begins.

Epic: none

Release note (sql change): when a user runs CREATE LOGICALLY REPLICATED TABLE,
they must specify one of the following options:
- UNIDIRECTIONAL: setups a unidirectional stream with fast initial scan
- BIDIRECTIONAL ON {dest uri}: sets up a bidirectional stream from the original
  dest to the original source.

msbutler commented:
First commit getting reviewed in #138680

jeffswenson commented on pkg/crosscluster/streamclient/partitioned_stream_client.go:
```diff
@@ -406,6 +406,26 @@ func (p *partitionedStreamClient) CreateForTables(
 	return spec, nil
 }
 
+func (p *partitionedStreamClient) SetupReverseStream(
```
I feel like `SetupReverseStream` is a bit of an odd man out compared to the other methods on the client. The other methods could be conceivably implemented over something like GRPC. `SetupReverseStream` on the other hand is accepting a sql command and is basically `conn.Exec` with a traced name.

I think this would match the pattern better if it took a struct that described the desired stream and internally constructed the statement. 

Alternatively, I think this should be named something that makes it clear what it really is: like `ExecStatement` or even expose the underlying pg connection object and run the statement on that.

msbutler replied:
This is a good callout. I'll refactor this to be `ExecStatement`.

I prefer that over your first suggestion as it's much easier to construct the reverse stream statement during the planning of the original ldr stmt, where we access to the OG parsed statement. Then, since this needs to be called during job execution, i merely need to persist the cmd instead of all the components of command.

jeffswenson commented on pkg/crosscluster/logical/create_logical_replication_stmt.go:
```diff
@@ -250,8 +255,23 @@ func createLogicalReplicationStreamPlanHook(
+		jobID := p.ExecCfg().JobRegistry.MakeJobID()
+		var reverseStreamCmd string
+		if stmt.CreateTable && options.BidirectionalURI() != "" {
+			// TODO: validate URI.
```
Validating the URI is a bit tricky. We really need the peer cluster to validate it.

msbutler replied:
>We really need the peer cluster to validate it.

Good point. I can plan to pass it as an extra arg to `CreateForTables`, so the og source can validate it. Sounds reasonable?

msbutler committed:
sql/externalcatalog: pass mutable flag to workaround resolver bug

For mysterious resons, the resolver cannot fetch offline descriptors unless the
mutable flag is passed.

Epic: none

Release note: none

jeffswenson reviewed (approved):
LGTM

msbutler commented:
TFTR!

bors r=jeffswenson
